### PR TITLE
lesson_2_decorator

### PR DESCRIPTION
**Pull request by crowstrong**
pull request

### CODE REVIEW

File: lesson_2_pull/lesson_2_decorator.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    def internal(f):
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            cache_key = (*args, tuple(kwargs.items()))
+            if cache_key in wrapper._cache:
+                wrapper._cache[cache_key]['freq'] += 1
+                return wrapper._cache[cache_key]['value']
+            result = f(*args, **kwargs)
+            if len(wrapper._cache) >= max_limit:
+                min_freq = min(wrapper._cache.values(), key=key_func)['freq']
```
Comment (mikolaz27) [inline lesson_2_pull/lesson_2_decorator.py:44]: remove key_func, you cant use external func in decorator, use lambda if you need function

File: lesson_2_pull/lesson_2_decorator.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                for k, v in wrapper._cache.items():
+                    if v['freq'] == min_freq:
+                        del wrapper._cache[k]
+                        break
+            wrapper._cache[cache_key] = {'value': result, 'freq': 1}
+            print('Cached url:')
+            for key in wrapper._cache.keys():
+                print(key[0])
```
Comment (mikolaz27) [inline lesson_2_pull/lesson_2_decorator.py:52]: remove prints

File: lesson_2_pull/lesson_2_decorator.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    @functools.wraps(count)
+    def wrapper(*args, **kwargs):
+        result = count(*args, **kwargs)
+        print(f'Function: {count.__name__}, use: {sys.getsizeof(result)} bytes')
+        return result
+
+    return wrapper
+
+
+def key_func(x):
```
Comment (mikolaz27) [inline lesson_2_pull/lesson_2_decorator.py:29]: remove

File: lesson_2_pull/lesson_2_decorator.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import functools
+import sys
+
+# from cachetools import cached, LFUCache
+from collections import OrderedDict
+
+import requests
+
+
+def object_memory(count):
```
Comment (mikolaz27) [inline lesson_2_pull/lesson_2_decorator.py:19]: why count? it should be function

File: lesson_2_pull/lesson_2_decorator.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    @functools.wraps(count)
+    def wrapper(*args, **kwargs):
+        result = count(*args, **kwargs)
+        print(f'Function: {count.__name__}, use: {sys.getsizeof(result)} bytes')
```
Comment (mikolaz27) [inline lesson_2_pull/lesson_2_decorator.py:23]: try to use tracemalloc lib for this

**Review by mikolaz27 — CHANGES_REQUESTED** [review]

**Review by crowstrong — COMMENTED** [review]
`
    return internal


@object_memory
# @cached(LFUCache(maxsize=2))
@lfu_cache_decorator(max_limit=2)
def fetch_url(url, first_n=100):
    """Fetch a given url"""
    res = requests.get(url)
    if first_n:
        return res.content[:first_n]
    else:
        return res.content


if __name__ == '__main__':
    fetch_url('https://www.google.com')
    fetch_url('https://www.google.com')
    fetch_url('https://www.google.com')
    fetch_url('https://www.google.com')
    fetch_url('https://www.youtube.com')
    fetch_url('https://ithillel.ua')
`

**Pull request merged**